.github/workflows/java.yml

```diff
@@ -26,46 +26,94 @@ on:
-            - name: Checkout
-              uses: actions/checkout@v4
+    check_should_run:
+      runs-on: ubuntu-latest
+      outputs:
+        should_run: ${{ steps.check_should_run.outputs.should_run }}
+      steps:
+        - name: Check for Core changes label
+          id: check_should_run
+          run: |
+            if [[ ("${{ github.event.label.name }}" == "Core changes" &&  "${{github.event_name}}" == 'pull_request_target') || "${{github.event_name}}" != 'pull_request_target' ]]; then
+              echo "should_run=true" >> $GITHUB_OUTPUT
+            else
+              echo "should_run=false" >> $GITHUB_OUTPUT
+            fi
 
-            - name: Load the engine matrix
-              id: load-engine-matrix
-              shell: bash
-              run: echo "matrix=$(jq -c . < .github/json_matrices/engine-matrix.json)" >> $GITHUB_OUTPUT
+    load-engine-matrix:
+      runs-on: ubuntu-latest
+      # We want to run this job only if the previous job outputs should_run=true,
+      # Which means that if the trigger was a label we run this job just if the label was Core changes
+      needs: check_should_run  
+      if: ${{ needs.check_should_run.outputs.should_run == 'true' }}   
+      outputs:
+        matrix: ${{ steps.load-engine-matrix.outputs.matrix }}
+      steps:
+          - name: Checkout
+            uses: actions/checkout@v4
+
+          - name: Load the engine matrix
+            id: load-engine-matrix
+            shell: bash
+            run: |
+              if [[ "${{ github.event_name }}" == "pull_request" || "${{ github.event_name }}" == "push" ]]; then
+                echo "matrix=$(jq -c '[.[] | select(.run == "always")]' < .github/json_matrices/engine-matrix.json)" >> $GITHUB_OUTPUT
+              else
+                echo "matrix=$(jq -c . < .github/json_matrices/engine-matrix.json)" >> $GITHUB_OUTPUT
+              fi
+
+    load-host-matrix:
+      runs-on: ubuntu-latest
+      needs: check_should_run
+      if: ${{ needs.check_should_run.outputs.should_run == 'true' }}
+      outputs:
+          matrix: ${{ steps.load-host-matrix.outputs.matrix }}
+      steps:
+          - name: Checkout
+            uses: actions/checkout@v4
+
+          - name: load host matrix
+            id: load-host-matrix
+            shell: bash
+            run: |
+              if [[ "${{ github.event_name }}" == "pull_request" || "${{ github.event_name }}" == "push" ]]; then
+                echo 'matrix={"include":' $(jq '[.[] | select(.OS == "ubuntu" and .RUNNER == "ubuntu-latest" and .TARGET == "x86_64-unknown-linux-gnu")]' .github/json_matrices/build-matrix.json) '}' >> $GITHUB_OUTPUT
+              else
+                echo 'matrix={"include":' $(jq '[.[] | select((.OS == "ubuntu" and .RUNNER == "ubuntu-latest" and .TARGET == "x86_64-unknown-linux-gnu") or (.OS == "macos" and .RUNNER == "macos-latest" and .TARGET == "aarch64-apple-darwin"))]' .github/json_matrices/build-matrix.json) '}' >> $GITHUB_OUTPUT
```

Comment:
Can you filter there for `.run == "always"` as you do for engine matrix? You can patch the matrix to keep logic in one place